cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,925 @@
+                               /*names=*/{"count(a)"},
+                               /*keys=*/{"b"}};
+  ARROW_ASSIGN_OR_RAISE(
+      cp::ExecNode * aggregate,
+      cp::MakeExecNode("aggregate", plan.get(), {source}, aggregate_options));
+
+  ARROW_RETURN_NOT_OK(
+      cp::MakeExecNode("sink", plan.get(), {aggregate}, cp::SinkNodeOptions{&sink_gen}));
+  auto schema = arrow::schema({
+      arrow::field("count(a)", arrow::int32()),
+      arrow::field("b", arrow::boolean()),
+  });
+
+  return ExecutePlanAndCollectAsTable(exec_context, plan, schema, sink_gen);
+}
+// (Doc section: Group Aggregate Example)
+
+// (Doc section: ConsumingSink Example)
+/**
+ * \brief
+ * Source-ConsumingSink
+ * This example shows how the data can be consumed within the execution plan
+ * by using a ConsumingSink node. There is no data output from this execution plan.
+ * \param exec_context : execution context
+ * \return arrow::Status
+ */
+arrow::Status SourceConsumingSinkExample(cp::ExecContext& exec_context) {
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<cp::ExecPlan> plan,
+                        cp::ExecPlan::Make(&exec_context));
+
+  ARROW_ASSIGN_OR_RAISE(auto basic_data, MakeBasicBatches());
+
+  auto source_node_options = cp::SourceNodeOptions{basic_data.schema, basic_data.gen()};
+
+  ARROW_ASSIGN_OR_RAISE(cp::ExecNode * source,
+                        cp::MakeExecNode("source", plan.get(), {}, source_node_options));
+
+  std::atomic<uint32_t> batches_seen{0};
+  arrow::Future<> finish = arrow::Future<>::Make();
+  struct CustomSinkNodeConsumer : public cp::SinkNodeConsumer {
+    CustomSinkNodeConsumer(std::atomic<uint32_t>* batches_seen, arrow::Future<> finish)
+        : batches_seen(batches_seen), finish(std::move(finish)) {}
+
+    arrow::Status Consume(cp::ExecBatch batch) override {
+      (*batches_seen)++;
+      return arrow::Status::OK();
+    }
+
+    arrow::Future<> Finish() override { return finish; }
+
+    std::atomic<uint32_t>* batches_seen;
+    arrow::Future<> finish;
+  };
+  std::shared_ptr<CustomSinkNodeConsumer> consumer =
+      std::make_shared<CustomSinkNodeConsumer>(&batches_seen, finish);
+
+  cp::ExecNode* consuming_sink;
+
+  ARROW_ASSIGN_OR_RAISE(consuming_sink,
+                        MakeExecNode("consuming_sink", plan.get(), {source},
+                                     cp::ConsumingSinkNodeOptions(consumer)));
+
+  ABORT_NOT_OK(consuming_sink->Validate());
+
+  ABORT_NOT_OK(plan->Validate());
+  std::cout << "Exec Plan created: " << plan->ToString() << std::endl;
+  // plan start producing
+  ARROW_RETURN_NOT_OK(plan->StartProducing());
+  // Source should finish fairly quickly
+  ARROW_RETURN_NOT_OK(source->finished().status());
+  std::cout << "Source Finished!" << std::endl;
+  // Mark consumption complete, plan should finish
+  arrow::Status finish_status;
+  // finish.Wait();
+  finish.MarkFinished(finish_status);
+  ARROW_RETURN_NOT_OK(plan->finished().status());
+  ARROW_RETURN_NOT_OK(finish_status);
```

Comment:
This line is redundant, finish_status is OK

Reply:
And actually, `finish.MarkFinished(Status::OK());` above is probably clearer since it's not necessarily obvious that a default-constructed Status is OK